src/main/java/com/azure/cosmos/kafka/connect/sink/BulkWriter.java

```diff
@@ -0,0 +1,146 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.cosmos.kafka.connect.sink;
+
+import com.azure.cosmos.BridgeInternal;
+import com.azure.cosmos.CosmosContainer;
+import com.azure.cosmos.CosmosException;
+import com.azure.cosmos.implementation.HttpConstants;
+import com.azure.cosmos.implementation.ImplementationBridgeHelpers;
+import com.azure.cosmos.implementation.routing.PartitionKeyInternal;
+import com.azure.cosmos.models.CosmosBulkItemResponse;
+import com.azure.cosmos.models.CosmosBulkOperationResponse;
+import com.azure.cosmos.models.CosmosBulkOperations;
+import com.azure.cosmos.models.CosmosItemOperation;
+import com.azure.cosmos.models.PartitionKey;
+import com.azure.cosmos.models.PartitionKeyDefinition;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.kafka.connect.sink.SinkRecord;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+import static com.azure.cosmos.implementation.guava25.base.Preconditions.checkArgument;
+import static com.azure.cosmos.implementation.guava25.base.Preconditions.checkNotNull;
+
+public class BulkWriter extends SinkWriterBase {
+    private static final Logger logger = LoggerFactory.getLogger(BulkWriter.class);
+
+    private final CosmosContainer cosmosContainer;
+    private final PartitionKeyDefinition partitionKeyDefinition;
+
+    public BulkWriter(CosmosContainer container, int maxRetryCount) {
+        super(maxRetryCount);
+
+        checkNotNull(container, "Argument 'container' can not be null");
+        this.cosmosContainer = container;
+        this.partitionKeyDefinition = container.read().getProperties().getPartitionKeyDefinition();
+    }
+
+    /***
+     * Bulk write the sink records.
+     *
+     * @param sinkRecords the sink records needs to be written.
+     * @return the list of sink write failed operations.
+     */
+    protected SinkWriteResponse writeCore(List<SinkRecord> sinkRecords) {
+
+        SinkWriteResponse sinkWriteResponse = new SinkWriteResponse();
+        if (sinkRecords == null || sinkRecords.isEmpty()) {
+            return sinkWriteResponse;
+        }
+
+
+        List<CosmosItemOperation> itemOperations = new ArrayList<>();
+        for (SinkRecord sinkRecord : sinkRecords) {
+            CosmosItemOperation cosmosItemOperation = CosmosBulkOperations.getUpsertItemOperation(
```

Comment:
Really always upsert? that makes it very inefficient when you know your data is immutable and you have to do retries based on some form of bookmarking. Upsert vs. insert should be configurable like in spark?

Reply:
yea make sense - the goal of this PR is to introduce bulk support by supporting the existing behavior(upsert always). And then adding support for different write strategy will happen next. Tracking item: https://github.com/microsoft/kafka-connect-cosmosdb/issues/487